pkg/controllers/resourceapply/resourceapply.go

```diff
@@ -158,6 +164,38 @@ func applyDeployment(ctx context.Context, client coreclientv1.Client, recorder r
 		return false, nil
 	}
 
+	// Check if deployment recreation needed
+	// Currently it is necessary if pod selector was changed
+	needRecreate := false
+	if !reflect.DeepEqual(existingCopy.Spec.Selector, required.Spec.Selector) {
+		needRecreate = true
+	}
+	if needRecreate {
+		klog.Infof("Deployment need to be recreated with new parameters")
+		// This check need for break a recursion in case if resource deletion was stuck during deletion
+		if existing.DeletionTimestamp != nil {
+			return false, fmt.Errorf("resource was already marked for deletion, returning")
+		}
```

Comment:
What if the object is coming from a cached client? Is this actually going to work/ could we see stale data and try deleting again?

What API response do you get if you delete an already deleted object?

Reply:
>What API response do you get if you delete an already deleted object?

Reply:
> 404

Is that also when the deleted object has a finalizer, I'm trying to work out if deleting an already "deleted" object that hasn't gone away throws an error

Reply:
I just checked, it doesn't throw an error, so we may loop here a few times still and have a successful "I deleted the resource" if the cache isn't updated yet <- this makes the recursion a possible issue and source of infinite recursion

Reply:
I removed recursion. If we can not rely to DeletionTimestamp, on a data which we obtaining from explicit Get invocation, if we have to suspect cache malfunction everywhere, ok, probably better to avoid somewhat hacky moves.